### PR TITLE
Subaddresses with multiple accounts support (preliminary)

### PR DESCRIPTION
**Pull request by kenshi84**
screen shot 2017-07-11 at 18 57 23
screen shot 2017-07-11 at 18 58 22

https://streamable.com/3nq9i

### CODE REVIEW

**Comment by snaggen** [comment]
Why is subadresses treated as separate accounts? Isn't it just different addresses to the same wallet? I would expect to have the wallet show the total amount, then maybe on the "account" page you could see the separate subtotals. On the recieve page I would suggest a drop down defaulting to a "main" adress (or maybe the last used). With the current version, the subadresses gets way to much focus.

**Comment by knaccc** [comment]
First, incredible work @kenshi84 !!!

From a UX perspective, I think there will be huge confusion as to what is going on, because it is not obvious from the UI that there are accounts, each with multiple receiving addresses.

In my opinion, there is a critical problem with the visual hierarchy. By having the account selection alongside operations which apply only to specific accounts, there is a disconnect between the logical hierarchy of operations and the visual hierarchy displayed in the GUI.

It needs to be obvious to a new user, without having to read a manual, that a Wallet contains Accounts, and each Account has multiple receiving addresses.

I've put together a mockup of two screens which I think make things clearer:

http://imgur.com/a/sGNcc   (home page / account selection screen)
http://imgur.com/a/4fHXq  (once account is selected)

**Comment by snaggen** [comment]
If subadresses are treated as separate accounts, then what is the benefit
from just using multiple wallets?

Den tis 11 juli 2017 18:14knaccc <notifications@github.com> skrev:

**Comment by knaccc** [comment]
@snaggen Multiple wallets mean multiplying the blockchain scanning time. If you had 5 independent wallets, it'd take you 5x as long to scan the blockchain, which is very computationally heavy (it's not just a bandwidth issue).

In contrast, you can have as many Accounts as you want, and the blockchain will be scanned just as quickly as if there were only one Account.

**Comment by snaggen** [comment]
I'm getting confused... 
You write that a wallet may have multiple accounts, and each account can have multiple subadresses... Please explain what you then mean with account, since I got the impression from the UI that subadresses are used as accounts. 

And as you write, and also as what is described in the Brief section of the subadresses mrl issue https://github.com/monero-project/research-lab/issues/7, one of the big reasons for subadresses is to be able to handout adresses to multiple places without them being linked, and also to use as a mechanism to know from where the money were sent. And I can see myself use subadresses for those two reasons, but I still don't see it as having different accounts. In my mind they are just different addresses to the same wallet. Then in the wallet it would be nice to have a section where I can see the wallets "sub-accounts" to keep track of each subadress and how much is recieved by that adress. 
So in my mind the wallet is the primary entity and the subadresses are a secondary entity. In the first UI it seems that each subadress is a primary entity, and the wallet total balance is only shown in the account tab.

**Comment by snaggen** [comment]
Also, if I come off as short and snarky it's just because my computer died and I'm currently doomed to use a phone to type all my comments. So, please give me some slack and read my comments as friendly and happy 😊. I'm really impressed by this work.

**Comment by knaccc** [comment]
@snaggen You're looking at an old document, the concept of accounts each with multiple receiving addresses is discussed here: https://github.com/monero-project/monero/pull/2056

Think of accounts as a way to organize your affairs. You can create an account for your main personal spending, an account that keeps track of the money you're making by buying and selling on online auctions, an account that you use to accept donations for an open source project, and an account that you're using for a side business you've started with your spouse.

Each account has a balance, and a primary subaddress. For each account, you can create further subaddresses.

Most people will probably just have one account, and will create a different address for each person they need to ask for payment from.

**Comment by snaggen** [comment]
Thanks for pointing me to the right document, now it makes more sense. Sorry for the noice.

**Comment by kenshi84** [comment]
@knaccc Thanks a lot for your ideas. I wish I had enough Qt programming skills, but all I could do was to read existing code, copy it and modify it so that I can provide a bare minimum proof of concept that demonstrates how to use the feature.

@Jaqueeee The current version is far from being polished with at least two issues that I'm aware of:

1. In the Receive tab, the original address text line is still there which is redundant because there is a list of subaddresses below it. I did so because the "copy to clipboard" button didn't seem working on my macOS.

2. Also in the Receive tab, the QR code image gets chopped off at the bottom which cannot be brought back even with scrolling. 

I really appreciate if you could step in to fix these problems and incorporate knaccc's suggestion, and also make sure that the layout doesn't get broken under mobile or other environments.

**Comment by Jaqueeee** [comment]
@kenshi Great work. No problem, I can do some polishing on this later. Haven't had time to test it yet.

**Comment by lethos3** [comment]
this looks amazing, the functionality that was missing to make me use the GUI

**Comment by Jaqueeee** [comment]
@kenshi84 @luigi1111 @fluffypony 
Since lots of wallet api changes was made in #2056 this PR is required for GUI to build/run. I think we should merge this ASAP and fix the graphical/ux glitches later on.

**Comment by Jaqueeee** [comment]
@kenshi84 I created a bunch of accounts and addresses. When i restarted the gui and reopened the wallet they were gone.

**Comment by kenshi84** [comment]
@Jaqueeee I did some research, and it looks like the problem is because of the wallet being unable to store when closing. I observe the same issue with the address book (i.e. any changes won't get saved to the file).

Perhaps this has something to do with one of the recently merged commits that changed the wallet behavior when closing? Paging @moneromooo-monero

**Comment by Jaqueeee** [comment]
@kenshi84 ok. That makes sense. Could you rebase this PR against master? There was a crash on exit bug that was fixed after you created this PR. Might be it.

**Comment by kenshi84** [comment]
@Jaqueeee Oh, sorry I forgot to rebase. Done

**Comment by kenshi84** [comment]
@Jaqueeee 
I'm confused about how I can build `libwallet_merged.a` from the master branch; if I simply do `./build.sh`, it builds from the release-v0.11 branch.

**Comment by Jaqueeee** [comment]
@kenshi84 thanks for the reminder. see #903

**Comment by kenshi84** [comment]
@Jaqueeee 

> @kenshi84 ok. That makes sense. Could you rebase this PR against master? There was a crash on exit bug that was fixed after you created this PR. Might be it.

I confirmed that the wallet can't save its state when exiting, which prevents the address book from being saved. I'm not sure how to fix it. CC @moneromooo-monero

**Comment by moneromooo-monero** [comment]
Sorry, I don't usually follow this repo so didn't see I was being asked something. In the wallet2 stuff, there's nohting that I know of which would prevent saving the address book, it's serialized in the cache with the rest of the data AFAICT.

**Comment by kenshi84** [comment]
I think #911 solves this store-on-exit issue which was introduced in https://github.com/monero-project/monero/pull/2289.

File: pages/Account.qml [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+            ListElement { address: "BZiySC"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+            ListElement { address: "BYqbdA"; label: ""; balance: "0"; unlockedBalance: "0" }
+        }
```
Comment (moneromooo-monero) [inline pages/Account.qml]: Is this meant to be still in ?

File: pages/Receive.qml [hunk 2]
```diff
@@ -216,6 +218,79 @@ Rectangle {
...
+                ListElement { address: "BfhF533yqmNfxfgoDXK1411DYveWMU1xRQzMPPpHVrZgeoCoVrQ9GCjY1mvDQQaoiCZYhKhGVHLkm51cgZYfDUpt5Bw4ZgB"; label: "ShapeShift purchase on 2017-Jul-09" }
+                ListElement { address: "BhTfiQY1Xj6AKXPUtDbunJPmqrbkha5qvNgYoipp6L7FQ8KTDciDWjF3YQ5g2dCLNLjJoUEJKVeYt9Rpb8tgDzT36fSoS9D"; label: "" }
+                ListElement { address: "BYoaHkHjFEY371Jq9fY9x9TJjaMeDXBYS9AfqCJ7SJoe5vuJLDPsyhU3RzDFNXKkEqBDqQMX1bDuUQbXKXctTGXY9ZgdGa1"; label: "ShapeShift purchase on 2017-Jul-11" }
+                ListElement { address: "BcxioZpKqYaZsSDHP6pmesaV7gMrPYxvHXsnkG5ceVGkasPR83cEJyvGNpckYvs2wP76HsM1KGo1mbEmmiburnCUT4aSEv1"; label: "ShapeShift purchase on 2017-Jul-13" }
+                ListElement { address: "BZzdxkqM8fJhZLATeyD3bd2cSrehGsANH4CgyJ4DBUamJAdyRTqkZ1W3kvN1UpzLz1iQU8XfEQUi15QyobFyMQVb3LDj4gZ"; label: "ShapeShift purchase on 2017-Jul-17" }
+                ListElement { address: "BYNUQQfEtg2gshraM8RPGvPUNbWcHHu33Tzhxy8ZrL7r7MYDkdPUCDFYxNawBNSAdjhKUCsiWjg5tQoReRMKGQ1fJS2QuEB"; label: "ShapeShift purchase on 2017-Jul-21" }
+            }
```
Comment (moneromooo-monero) [inline pages/Receive.qml]: Is this supposed to be still in ?

File: src/libwalletqt/PendingTransaction.cpp [hunk 3]
```diff
@@ -50,6 +50,16 @@ quint64 PendingTransaction::txCount() const
...
 
+QList<QVariant> PendingTransaction::subaddrIndices() const
+{
+    std::vector<std::set<uint32_t>> subaddrIndices = m_pimpl->subaddrIndices();
+    QList<QVariant> result;
+    for (const auto& x : subaddrIndices)
+        for (uint32_t i : x)
+            result.push_back(i);
+    return result;
+}
```
Comment (moneromooo-monero) [inline src/libwalletqt/PendingTransaction.cpp]: This looks like you'll get duplicates when having several txes. Maybe removing any would be better.

File: src/libwalletqt/Subaddress.h [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+#include <QDateTime>
+
+class Subaddress : public QObject
+{
+    Q_OBJECT
+public:
+    Q_INVOKABLE QList<Monero::SubaddressRow*> getAll(bool update = false) const;
+    Q_INVOKABLE Monero::SubaddressRow * getRow(int index) const;
+    Q_INVOKABLE void addRow(quint32 accountIndex, const QString &label) const;
+    Q_INVOKABLE void setLabel(quint32 accountIndex, quint32 addressIndex, const QString &label) const;
```
Comment (moneromooo-monero) [inline src/libwalletqt/Subaddress.h]: setters should not be const (might explain the mutable below)

File: src/libwalletqt/Wallet.cpp [hunk 5]
```diff
@@ -458,6 +516,34 @@ AddressBookModel *Wallet::addressBookModel() const
     return m_addressBookModel;
 }
 
+Subaddress *Wallet::subaddress() const
```
Comment (moneromooo-monero) [inline src/libwalletqt/Wallet.cpp]: const methods should return const pointers to their data.

**Comment by kayront** [comment]
See bottom/final part of #1864 for my comments on why this is a very worthy addition.

In my opinion it would make sense to prioritize this feature - it enables more advanced use cases and because of the base layer privacy, it functions without damaging privacy for the end-user.

**Comment by xmrdsc** [comment]
Implemented via #1905 

+resolved

**Pull request closed without merging**